Style guide (#34)
canonical/ubuntu-packaging-guide: docs/reference/style-guide.rst

> Ubuntu Package Guide -- Style Guide

Reviewer: ```suggestion
Ubuntu Package Guide style guide
```

===================================

.. note::

>     This Style Guide is based on `Canonical' reStructuredText style guide`_.

Reviewer: ```suggestion
    This style guide is based on the `Canonical reStructuredText style guide`_.
```

All documentation files should use `reStructuredText`_ (reST) syntax. Community
contributions in `markdown`_ are welcome, but they should be reformated to reST, before

> they are merged into the main branch 

Reviewer: ```suggestion
they are merged into the main branch.
```

Reviewer: Is this on the maintainers or the contributors to do this?

> For general style conventions, see the `Canonical Documentation Style Guide`_.

Reviewer: ```suggestion
For general style conventions, see the `Canonical documentation style guide`_.
```

Headings
--------

.. grid:: 2
    
    .. grid-item::

        .. code:: reStructuredText

            Title
            =====

    .. grid-item::
        :child-align: center

>         Page title and ``<h1>`` heading

Reviewer: Might be worth also noting there should be only one of these per page

    .. grid-item::

        .. code:: reStructuredText

            Heading
            -------   
            
    .. grid-item::
        :child-align: center

        ``<h2>`` heading
    
    .. grid-item::

        .. code:: reStructuredText

            Heading
            ~~~~~~~   
            
    .. grid-item::
        :child-align: center

        ``<h3>`` heading

    .. grid-item::

        .. code:: reStructuredText

            Heading
            ^^^^^^^
            
    .. grid-item::
        :child-align: center

        ``<h4>`` heading        

    .. grid-item::

        .. code:: reStructuredText

            Heading
            .......   
            

    .. grid-item::
        :child-align: center

        ``<h5>`` heading

> Underlines must be the same length as the title or heading.

Reviewer: They can also be longer, just not shorter

Reviewer: (it's fine if we want to enforce them being the same length, though!)

Adhere to the following conventions:

- Do not use consecutive headings without intervening text.
- Be consistent with the characters you use for each level. Use the ones specified above.
- Do not skip levels (for example, do not follow an H2 heading with an H4 heading).

> - Use sentence style for headings (capitalise only the first word).

Reviewer: I'm not a fan of the "negative" phrasing employed in the original, and would be happy if we avoided it here. For example, instead of saying "Do not use consecutive headings without intervening text" we could use "Make sure that you include text between consecutive headings" and so on. It sounds less stark and a bit more welcoming to phrase it in a more positive way of "you should do this" vs. "don't do that".

Inline formatting
-----------------

.. list-table::
   :header-rows: 1

   * - Input
     - Output
   * - ``*Italic*``
     - *Italic*
   * - ``**Bold**``
     - **Bold**
   * - ````code````
     - ``code``
   * - ``:term:`Ubuntu```
     - :term:`Ubuntu`
   * - ``:file:`/file/path```
     - :file:`/file/path`
   * - ``:manpage:`man(1)```
     - :manpage:`man(1)`
   * - ``:command:`command```
     - :command:`command`
   * - ``:kbd:`key```
     - :kbd:`key`

Adhere to the following conventions:

> - Use italics sparingly. Common uses for italics are titles and names (for example,

Reviewer: The Ubuntu font, being so narrow, makes italics really hard to read for people with dyslexia (who already struggle with italics). I recommend we diverge from the original again here and suggest using bold for emphasis instead of italic (in point one) and that italic should be avoided through rewriting (point two).

  when referring to a section title that you cannot link to, or when introducing
  the name for a concept).
- Use bold sparingly. Avoid using bold for emphasis and rather rewrite the sentence
  to get your point across.

> Notes

Reviewer: I believe that these notes (as a collective) should be called "admonitions"

-----

.. grid:: 2
    :gutter: 2
    
    .. grid-item::
        :child-align: center

        .. code:: reStructuredText

            .. note::
                A note.

    .. grid-item-card::

            .. note::
                A note.

    .. grid-item::
        :child-align: center

        .. code:: reStructuredText

            .. tip::
                A tip.

    .. grid-item-card::

            .. tip::
                A tip. 
    
    .. grid-item::
        :child-align: center

        .. code:: reStructuredText

            .. important::
                Important information.

    .. grid-item-card::

            .. important::
                Important information. 
    
    .. grid-item::
        :child-align: center

        .. code:: reStructuredText

            .. caution::
                This might damage your hardware!

    .. grid-item-card::

            .. caution::
                This might damage your hardware!

Adhere to the following conventions:

- Use notes sparingly.

> - Only use the following note types: ``note``, ``tip``, ``important``, ``caution``

Reviewer: This point is under discussion currently and seems to be largely a matter of taste more than anything else - there's certainly no technical reason someone can't use the rest of the range of admonitions available, the only consensus is that they should be used sparingly.

- Only use a caution if there is a clear hazard of hardware damage or data loss.

Code blocks
-----------

To start a code block, start a code block with ``.. code::``.

> The code block must be surrounded by empty lines.

Reviewer: And there must be an empty line between the ``code`` directive and the content of the code block

When starting a code block, you can specify the code language to enforce a specific
sytax highlighting, but in many cases, the default sytax highlighting works just fine.

> 

Reviewer: There should be a mention here that the contents of the code block are determined by the indentation level. Anything that's indented will be included in the code block, even if there are empty lines between the content.

                echo "Hello, World!"
     - .. code::
       
           echo "Hello, World!"
   * - .. code:: reStructuredText

            .. code:: text
       
                echo "Hello, World!"
     - .. code:: text
       
           echo "Hello, World!"
   * - .. code:: reStructuredText

            .. code:: bash
       
                echo "Hello, World!"
     - .. code:: bash
       
           echo "Hello, World!"

Fore the sake of consistency, please **do not** use the pattern:

.. list-table::
   :header-rows: 1

   * - Input
     - Output
   * - .. code:: reStructuredText
       
            Look, after this text starts a code block::
       
                echo "Hello, World!"
     - Look, after this text starts a code block::
       
           echo "Hello, World!"

Instead you can do this:

.. list-table::
   :header-rows: 1

   * - Input
     - Output
   * - .. code:: reStructuredText
       
            Look, after this text starts a code block:

            .. code::
       
                echo "Hello, World!"
     - Look, after this text starts a code block:

       .. code::
       
           echo "Hello, World!"

Definition lists
----------------

> 

Reviewer: When and where should these be used? (same question for cards and tabs)

       
                    Content 2
       
           .. tab-set::
       
               .. tab-item:: Label1
                   :sync: key1
       
                   Content 1
       
               .. tab-item:: Label2
                   :sync: key2
       
                   Content 2
     - .. tab-set::

           .. tab-item:: Label1
               :sync: key1
   
               Content 1
   
           .. tab-item:: Label2
               :sync: key2
   
               Content 2
   
       .. tab-set::
   
           .. tab-item:: Label1
               :sync: key1
   
               Content 1
   
           .. tab-item:: Label2
               :sync: key2
   
               Content 2

More useful markup
------------------

.. list-table::
   :header-rows: 1

   * - Input
     - Output
   * - .. code:: reStructuredText

           | Line 1
           | Line 2
           | Line 3
     - | Line 1
       | Line 2
       | Line 3
   * - .. code:: reStructuredText

           ----
     - A horizontal line
     
       (for technical reasons the horizontal line can not be displayed here)
   * - .. code:: reStructuredText

           .. This is a comment
     - 

> 

Reviewer: We should probably include link formatting and options as well. First, we need to decide if we want to go ahead with translations, because that will have to factor into our decisions about how links should be treated (doing them the standard way as presented in the reST style guide that was the basis of this doc won't work very well and will make translators' jobs harder).

> .. _Canonical' reStructuredText style guide: https://canonical-documentation-with-sphinx-and-readthedocscom.readthedocs-hosted.com/style-guide/

Reviewer: If you accepted my suggestion in the text, this text will need to be changed to match or it won't render

.. _reStructuredText: https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html
.. _markdown: https://www.sphinx-doc.org/en/master/usage/markdown.html

> .. _Canonical Documentation Style Guide: https://docs.ubuntu.com/styleguide/en

Reviewer: Same as line 389
